core-foundation/src/propertylist.rs

```diff
@@ -190,14 +190,18 @@ impl CFPropertyList {
     /// // Cast it up to a property list.
     /// let propertylist: CFPropertyList = string.to_CFPropertyList();
     /// // Cast it down again.
-    /// assert!(propertylist.downcast::<_, CFString>().unwrap().to_string() == "FooBar");
+    /// assert!(propertylist.downcast::<CFString>().unwrap().to_string() == "FooBar");
     /// ```
     ///
     /// [`CFPropertyList`]: struct.CFPropertyList.html
     /// [`Box::downcast`]: https://doc.rust-lang.org/std/boxed/struct.Box.html#method.downcast
-    pub fn downcast<Raw, T: CFPropertyListSubClass<Raw>>(&self) -> Option<T> {
-        if self.instance_of::<_, T>() {
-            Some(unsafe { T::wrap_under_get_rule(self.0 as *const Raw) })
+    pub fn downcast<T: CFPropertyListSubClass>(&self) -> Option<T> {
+        if self.instance_of::<T>() {
+            let ptr = &self.0 as *const _ as *const <T as TCFType>::TypeRef;
+            unsafe {
+                let reference: &<T as TCFType>::TypeRef = &*ptr;
+                Some(T::wrap_under_get_rule(mem::transmute_copy(reference)))
```

Comment:
This implementation became uglier in order to get rid of the `Raw` generic. Using `mem::transmute_copy` is not very recommended maybe, but I hope this is used correctly. Here `reference` is of type `&CFFooRef`, which after the `transmute_copy` should be a `CFFooRef`. So the `transmute_copy` don't change the type, it just helps me take ownership of a dereferenced raw pointer. Which should be safe here unless I'm mistaken.

Reply:
Maybe a more elegant solution is possible? I would be eager to learn a better way if there is one.

Reply:
Help me understand why we can't use `Some(T::wrap_under_get_rule(ptr))`? That method should accept `*const T::ConcreteTypeRef` if I'm not mistaken.

Reply:
This is of course depending on the assumption that all `TCFType::TypeRef`s have the same size as `*const c_void`. This is true for all the possible subclasses of `CFPropertyList` at the moment and probably forever, but nothing is enforcing it.

Reply:
If I add a second associated type, `TCFType::RawTypeRef` (or similar name), that we set to the corresponding `__CFFoo` type. Then I can refer to `*const Self::RawTypeRef` in places and do proper casts. From some early tinkering around it looks like that can get rid of all these strange solutions.